### ISSUE TITLE
Does oc.dict.values preserve order?

### ISSUE THREAD

**Issue by will-gp** [omry/omegaconf#982]
I am wondering if `oc.dict.values` preserves order.

Given the following example:

pets.yaml
```
zebra:
    - name: Bob
    - age: 24
cat:
    - name: Louis
    - age: 5
```

and another yaml file which uses `pets.yaml`:
```
defaults:
  - pets

profile:
   - ${oc.dict.values:pets}
```

will `oc.dict.values` preserve the order of the objects in `pets.yaml`?
ie. the order will **always** be the values of [zebra,cat] and never [cat, zebra]

**Comment by Jasha10** [comment]
TLDR: Yes, `oc.dict.{keys,values}` iterate over keys/values in order (for python version >= 3.7)

---

`oc.dict.keys` and `oc.dict.values` iterate over the keys/values in the same order that they appear in the `DictConfig` object. (You can see the implementation of `oc.dict.keys` and `oc.dict.values` in [this file](https://github.com/omry/omegaconf/blob/7e519172b26fa7e7b147a8acb88c4951805a982d/omegaconf/resolvers/oc/dict.py)) Under the hood, `DictConfig` uses a python dictionary to store its contents (i.e. its keys and values). For Python version >=3.7, python dictionaries are guaranteed to preserve order of keys/values. This means that, for python >= 3.7, `DictConfig` itself is stable in terms of the order of its keys, and `oc.dict.{keys,values}` will preserve that order.

I noticed that your example above uses yaml files and a hydra-like defaults list. Note that according to the [yaml spec](https://yaml.org/spec/1.2.2/#3221-mapping-key-order), yaml mappings (i.e. "yaml dicts") are technically *unordered*. OmegaConf and Hydra use the [`pyyaml`](https://github.com/yaml/pyyaml) library to handle yaml parsing, and I'm not sure if `pyyaml` guarantees that the keys/values from yaml files will be parsed in order.

**Comment by will-gp** [comment]
Thanks @Jasha10. It does seem, based on a number of discussions, that `pyyaml` preserves insertion order. In this case, even though the yaml mappings would be ordered, this would still represent a valid yaml.

I've included the link(s) below for reference:
[https://github.com/yaml/pyyaml/issues/110](https://github.com/yaml/pyyaml/issues/110)
[https://github.com/yaml/pyyaml/issues/523](https://github.com/yaml/pyyaml/issues/523)

I would note that although loading the yaml preserves order, `yaml.dump()` **does not**, which is a source of confusion for many.
